### PR TITLE
Assemble P2SH deposit transaction

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/tbtc-v2/issues/79

This pull request introduces tooling meant to be used for assembling and broadcasting Bitcoin P2SH deposit transactions. Specifically, it extends the `TBTC` interface by adding the following methods:
- `makeDeposit`: Function which can be used to perform an end-to-end deposit flow. Takes deposit parameters, depositor private key, and Bitcoin client as arguments. It scans the depositor account for UTXOs, constructs the P2SH transaction according to given parameters, and broadcasts the transaction over the network. The main purpose of that function are future e2e tests of the whole system.
- `createDepositTransaction`: Function which assembles the P2SH transaction based on given deposit parameters, depositor private key, and UTXOs. It returns the transaction in a raw format. Such a transaction can be used for testing or can be broadcast out-of-band.
- `createDepositScript`: Creates the Bitcoin script used as a lock of the P2SH deposit transaction output. It implements the script described in [RFC-1](https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-1.adoc#23-depositing). Returns the serialized script in a raw format for further processing.
- `createDepositScriptHash`:  Function that computes the hash of the script returned by `createDepositScript`. This hash has several use cases. It is used during P2SH transaction creation and while deriving deposit target address.
- `createDepositAddress`: Function that computes the deposit target address for the given Bitcoin network using the output of `createDepositScriptHash`. That address can be passed to the depositor who wants to perform the Bitcoin P2SH transaction on their own.

### CODE REVIEW

**Comment by pdyraga** [comment]
What if we rename `tbtc-ts` to `typescript`? We have `solidity` and TS code is a part of `tbtc-v2`, it does not have to be a separate module. We can do it later, once PRs are merged so that we do not run into conflicts.

File: tbtc-ts/src/deposit.ts [hunk 1]
```diff
@@ -1,7 +1,248 @@
...
+  const depositorKeyRing = createKeyRing(depositorPrivateKey)
+  const depositorAddress = depositorKeyRing.getAddress("string")
+
+  const utxos = await bitcoinClient.findAllUnspentTransactionOutputs(
+    depositorAddress
+  )
+
+  const utxosWithRaw: (UnspentTransactionOutput & RawTransaction)[] = []
+  for (const utxo of utxos) {
+    const rawTransaction = await bitcoinClient.getRawTransaction(
```
Comment (tomaszslabon) [inline tbtc-ts/src/deposit.ts:69]: We will have to add error handling for communication with bitcoin, as it may always fail (I suggest we do that in another PR).

**Review by tomaszslabon — APPROVED** [review]

File: tbtc-ts/src/deposit.ts (hunk 1 above)
Comment (lukasz-zimnoch) [reply tbtc-ts/src/deposit.ts:69]: Yep for sure but that's the matter of the implementation, not the `BitcoinClient` interface. Agree we should address it within the PR introducing Electrum client.

**Comment by lukasz-zimnoch** [comment]
> What if we rename `tbtc-ts` to `typescript`? We have `solidity` and TS code is a part of `tbtc-v2`, it does not have to be a separate module. We can do it later, once PRs are merged so that we do not run into conflicts.

Personally, I'm not a fan of naming those subdirectories by language names but I'd rather name them feature-wise. In my opinion, names like `solidity` or `typescript` give no clue about what that directory contains and should contain. For example, `solidity` directory contains typescript tests and deployment scripts as well. Such a directories also tend to be just bags for code written in a given language, not necessarily connected functionally. However, I know it's the convention we used in our projects so far and that requires more discussion to be changed. I totally agree we should address it in a separate PR/issue once we merge some code and see the direction things actually go.
